### PR TITLE
Add Vulntotal CLI

### PR DESCRIPTION
**Pull request by keshav-space**
``` py
❯ python vulntotal/vulntotal_cli.py --help

Usage: vulntotal_cli.py [OPTIONS] [PURL]

  Runs the PURL through all the available  DataSources and group vulnerability
  by CVEs. Use the special '-' file name to print JSON or YAML results on
  screen/stdout.

Options:
  -l, --list   Lists all the available DataSources.
  --json FILE  Write output as pretty-printed JSON to FILE.
  --yaml FILE  Write output as YAML to FILE.
  -h, --help   Show this message and exit.
```

</br>
</br>
Along with some advanced hidden options especially useful for debugging and development.

``` py

Options:
  -e, --enable                    Enable these datasource/s only.
  -d, --disable                   Disable these datasource/s.
  --ecosystem                     Lists ecosystem supported by active DataSources
  --raw                           List of all the raw response from DataSources.
  --no-threading                  Run DataSources sequentially.
  -p, --pagination                Enable default pagination.
  --no-group                      Don't group by CVE.
```

### CODE REVIEW

**Comment by keshav-space** [comment]
``` py
❯ python vulntotal/vulntotal-cli.py -e deps osv -p 'pkg:pypi/jinja2@2.4.1'
PURL:  pkg:pypi/jinja2@2.4.1
Active Validators:  deps, osv
+-----------+----------------------+---------------------+---------------------+
| VALIDATOR |       ALIASES        |      AFFECTED       |        FIXED        |
+===========+======================+=====================+=====================+
|   DEPS    | CVE-2020-28493       | 2.0  2.0rc1  2.1    | 2.11.3  3.0.0       |
|           |                      | 2.1.1  2.10  2.10.1 | 3.0.0a1  3.0.0rc1   |
|           |                      | 2.10.2  2.10.3      | 3.0.0rc2  3.0.1     |
|           |                      | 2.11.0  2.11.1      | 3.0.2  3.0.3  3.1.0 |
|           |                      | 2.11.2  2.2  2.2.1  | 3.1.1  3.1.2        |
|           |                      | 2.3  2.3.1  2.4     |                     |
|           |                      | 2.4.1  2.5  2.5.1   |                     |
|           |                      | 2.5.2  2.5.3  2.5.4 |                     |
|           |                      | 2.5.5  2.6  2.7     |                     |
|           |                      | 2.7.1  2.7.2  2.7.3 |                     |
|           |                      | 2.8  2.8.1  2.9     |                     |
|           |                      | 2.9.1  2.9.2  2.9.3 |                     |
|           |                      | 2.9.4  2.9.5  2.9.6 |                     |
+-----------+----------------------+---------------------+---------------------+
|    OSV    | CVE-2014-1402        | 0  2.0  2.0rc1  2.1 | 2.7.2               |
|           | PYSEC-2014-8         | 2.1.1  2.2  2.2.1   |                     |
|           |                      | 2.3  2.3.1  2.4     |                     |
|           |                      | 2.4.1  2.5  2.5.1   |                     |
|           |                      | 2.5.2  2.5.3  2.5.4 |                     |
|           |                      | 2.5.5  2.6  2.7     |                     |
|           |                      | 2.7.1               |                     |
+-----------+----------------------+---------------------+---------------------+
|    OSV    | CVE-2014-0012        | 0  2.0  2.0rc1  2.1 | acb672b6a179567632e |
|           | PYSEC-2014-82        | 2.1.1  2.2  2.2.1   | 032f547582f30fa2f4a |
|           |                      | 2.3  2.3.1  2.4     | a7                  |
|           |                      | 2.4.1  2.5  2.5.1   |                     |
|           |                      | 2.5.2  2.5.3  2.5.4 |                     |
|           |                      | 2.5.5  2.6  2.7     |                     |
|           |                      | 2.7.1  2.7.2        |                     |
+-----------+----------------------+---------------------+---------------------+
|    OSV    | CVE-2019-10906       | 0  2.0  2.0rc1  2.1 | 2.10.1              |
|           | GHSA-462w-v97r-4m45  | 2.1.1  2.10  2.2    |                     |
|           | PYSEC-2019-217       | 2.2.1  2.3  2.3.1   |                     |
|           |                      | 2.4  2.4.1  2.5     |                     |
|           |                      | 2.5.1  2.5.2  2.5.3 |                     |
|           |                      | 2.5.4  2.5.5  2.6   |                     |
|           |                      | 2.7  2.7.1  2.7.2   |                     |
|           |                      | 2.7.3  2.8  2.8.1   |                     |
|           |                      | 2.9  2.9.1  2.9.2   |                     |
|           |                      | 2.9.3  2.9.4  2.9.5 |                     |
|           |                      | 2.9.6               |                     |
+-----------+----------------------+---------------------+---------------------+
|    OSV    | CVE-2016-10745       | 0  2.0  2.0rc1  2.1 | 9b53045c34e61013dc8 |
|           | GHSA-hj2j-77xm-mc5v  | 2.1.1  2.2  2.2.1   | f09b7e52a555fa16bed |
|           | PYSEC-2019-220       | 2.3  2.3.1  2.4     | 16                  |
|           |                      | 2.4.1  2.5  2.5.1   |                     |
|           |                      | 2.5.2  2.5.3  2.5.4 |                     |
|           |                      | 2.5.5  2.6  2.7     |                     |
|           |                      | 2.7.1  2.7.2  2.7.3 |                     |
|           |                      | 2.8                 |                     |
+-----------+----------------------+---------------------+---------------------+
|    OSV    | CVE-2020-28493       | 0  2.0  2.0rc1  2.1 | 2.11.3              |
|           | GHSA-g3rq-g295-4j3m  | 2.1.1  2.10  2.10.1 |                     |
|           | PYSEC-2021-66        | 2.10.2  2.10.3      |                     |
|           | SNYK-PYTHON-         | 2.11.0  2.11.1      |                     |
|           | JINJA2-1012994       | 2.11.2  2.2  2.2.1  |                     |
|           |                      | 2.3  2.3.1  2.4     |                     |
|           |                      | 2.4.1  2.5  2.5.1   |                     |
|           |                      | 2.5.2  2.5.3  2.5.4 |                     |
|           |                      | 2.5.5  2.6  2.7     |                     |
|           |                      | 2.7.1  2.7.2  2.7.3 |                     |
|           |                      | 2.8  2.8.1  2.9     |                     |
|           |                      | 2.9.1  2.9.2  2.9.3 |                     |
|           |                      | 2.9.4  2.9.5  2.9.6 |                     |
+-----------+----------------------+---------------------+---------------------+
(END)
```

**Comment by keshav-space** [comment]
Now CLI also supports the listing of `supported ecosystems`

```
❯ python vulntotal/vulntotal-cli.py -le
Active Validators: DEPS, GITHUB, OSV

Ecosystem supported by active validators
ANDROID
CARGO
COMPOSER
CRATES.IO
DEBIAN
ERLANG
GEM
GO
HEX
LINUX
MAVEN
NPM
NUGET
OSS-FUZZ
PACKAGIST
PYPI
RUBYGEMS
RUST
```

File: vulntotal/vulntotal-cli.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+Description:
+    When no -e/-d flag is provided CLI will run the PURL through
+    all the available validators.
+
+
+Optional arguments:
+  -h, --help                        Show this help message and exit
+  -l, --list                        Lists all the available validators
+  -le, --ecosystem                  Lists ecosystem supported by active vendor
```
Comment (pombredanne) [inline vulntotal/vulntotal-cli.py:55]: single dash options should be single letter only.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! LGTM very nice... Do you mind to use click may be rather than argparse? Also I think there should be only two user-visible options: `--purl` and `--list` all the other options are debugging options. The user should not be able routinely to select/enable/disable validators at all. This means that we have not done the right job.

**Comment by pombredanne** [comment]
IMHO we should also find a way to report some JSON output.

**Comment by pombredanne** [comment]
Some example of CLI could be https://github.com/nexB/python-inspector/blob/main/src/python_inspector/resolve_cli.py

**Comment by TG1999** [comment]
Add doctsrings for all the functions and also add doctests/ unit tests for functions which have not been tested at all

**Comment by armijnhemel** [comment]
> ```python
> | CVE-2020-28493 | VULNERABLECODE | CVE-2020-28493 | 2.10.1-3       | 1.11.3-r0     |
> |                |                | GHSA-g3rq-g295 | 2.10.3-6       | 1.11.3-r0     |
> |                |                | -4j3m          | 2.7.0-12       | 1.11.3-r0     |
> |                |                |                | 2.7.18-3       | 1.11.3-r0     |
> ```

That looks like an odd result as I don't think that version `1.11.3-r0` actually exists. Maybe a bug in the VulnerableCode data? I would have expected to see `2.11.3-r0`

**Comment by TG1999** [comment]
> > ```python
> > | CVE-2020-28493 | VULNERABLECODE | CVE-2020-28493 | 2.10.1-3       | 1.11.3-r0     |
> > |                |                | GHSA-g3rq-g295 | 2.10.3-6       | 1.11.3-r0     |
> > |                |                | -4j3m          | 2.7.0-12       | 1.11.3-r0     |
> > |                |                |                | 2.7.18-3       | 1.11.3-r0     |
> > ```
> 
> That looks like an odd result as I don't think that version `1.11.3-r0` actually exists. Maybe a bug in the VulnerableCode data? I would have expected to see `2.11.3-r0`

@armijnhemel it's not a bug in Vulnerablecode, the version `1.11.3-r0` is not from pypi ecosystem but from apline ecosystem `pkg:alpine/py3-jinja2@1.11.3-r0?arch=aarch64&distroversion=edge&reponame=main` vulntotal needs some sort of filtering to filter out the purls that are related to the ecosystem of the requested purl.

**Comment by armijnhemel** [comment]
> > > ```python
> > > | CVE-2020-28493 | VULNERABLECODE | CVE-2020-28493 | 2.10.1-3       | 1.11.3-r0     |
> > > |                |                | GHSA-g3rq-g295 | 2.10.3-6       | 1.11.3-r0     |
> > > |                |                | -4j3m          | 2.7.0-12       | 1.11.3-r0     |
> > > |                |                |                | 2.7.18-3       | 1.11.3-r0     |
> > > ```
> > 
> > 
> > That looks like an odd result as I don't think that version `1.11.3-r0` actually exists. Maybe a bug in the VulnerableCode data? I would have expected to see `2.11.3-r0`
> 
> @armijnhemel it's not a bug in Vulnerablecode, the version `1.11.3-r0` is not from pypi ecosystem but from apline ecosystem `pkg:alpine/py3-jinja2@1.11.3-r0?arch=aarch64&distroversion=edge&reponame=main` vulntotal needs some sort of filtering to filter out the purls that are related to the ecosystem of the requested purl.
> 
> cc @keshav-space

I will open a different issue, as I have found what the bug is.

**Comment by keshav-space** [comment]
> @armijnhemel it's not a bug in Vulnerablecode, the version `1.11.3-r0` is not from pypi ecosystem but from apline ecosystem `pkg:alpine/py3-jinja2@1.11.3-r0?arch=aarch64&distroversion=edge&reponame=main` vulntotal needs some sort of filtering to filter out the purls that are related to the ecosystem of the requested purl.
> 
> cc @keshav-space

VulnTotal is making this request.

```python
response = requests.post(
    "http://localhost:8001/api/packages/bulk_search/",
    json={"purls": ["pkg:pypi/jinja2@2.4.1"]},
)


```


The point here is that if I'm making an explicit request for `pypi` ecosystem, why should I be getting anything from the alpine?

**Comment by TG1999** [comment]
> > @armijnhemel it's not a bug in Vulnerablecode, the version `1.11.3-r0` is not from pypi ecosystem but from apline ecosystem `pkg:alpine/py3-jinja2@1.11.3-r0?arch=aarch64&distroversion=edge&reponame=main` vulntotal needs some sort of filtering to filter out the purls that are related to the ecosystem of the requested purl.
> > cc @keshav-space
> 
> VulnTotal is making this request.
> 
> ```python
> response = requests.post(
>     "http://localhost:8001/api/packages/bulk_search/",
>     json={"purls": ["pkg:pypi/jinja2@2.4.1"]},
> )
> ```
> 
> The point here is that if I'm making an explicit request for `pypi` ecosystem, why should I be getting anything from the alpine?

@keshav-space which branch of vulnerablecode are you using in your local checkout? if you use the latest branch you will only get purls of `pypi` ecosystem

**Comment by armijnhemel** [comment]
> > > > ```python
> > > > | CVE-2020-28493 | VULNERABLECODE | CVE-2020-28493 | 2.10.1-3       | 1.11.3-r0     |
> > > > |                |                | GHSA-g3rq-g295 | 2.10.3-6       | 1.11.3-r0     |
> > > > |                |                | -4j3m          | 2.7.0-12       | 1.11.3-r0     |
> > > > |                |                |                | 2.7.18-3       | 1.11.3-r0     |
> > > > ```
> > > 
> > > 
> > > That looks like an odd result as I don't think that version `1.11.3-r0` actually exists. Maybe a bug in the VulnerableCode data? I would have expected to see `2.11.3-r0`
> > 
> > 
> > @armijnhemel it's not a bug in Vulnerablecode, the version `1.11.3-r0` is not from pypi ecosystem but from apline ecosystem `pkg:alpine/py3-jinja2@1.11.3-r0?arch=aarch64&distroversion=edge&reponame=main` vulntotal needs some sort of filtering to filter out the purls that are related to the ecosystem of the requested purl.
> > cc @keshav-space
> 
> I will open a different issue, as I have found what the bug is.

It seems that what I found is indeed a different issue than what is currently being discussed ;-)

I have opened a separate bug report in #915

**Comment by keshav-space** [comment]
> @keshav-space which branch of vulnerablecode are you using in your local checkout? if you use the latest branch you will only get purls of `pypi` ecosystem

Okay, let me try the latest branch.

**Pull request merged**